Expose feedback cache API for Feedback 2.0 UX

JunDai opened the pull request:
### Description
<!--
Include issue references (e.g., fixes [#issue](link))
Include necessary implementation details (e.g. I opted to use this algorithm because ... and test it in this way ...).
-->

The Feedback 2.0 flow allows to add more information on arrival. To support the new feedback UX change, the Core SDK should provide these APIs:
* cache user feedback
* retrieve cached feedback list
* post cached feedbacks

### Changelog
<!--
Include changelog entry (e.g. Fixed an unexpected change in recenter button when resuming the app.).
See https://github.com/mapbox/navigation-sdks/blob/main/documentation/android-changelog-guidelines.md.
You can remove the changelog block and add a `skip changelog` label, when applicable.
 -->
```
<changelog>Expose APIs to support caching user feedback events.</changelog>
```

<!--
---------- CHECKLIST ----------
1. Add related labels (`bug`, `feature`, `new API(s)`, `SEMVER-MAJOR`, `needs-backporting`, etc.).
2. Update progress status on the project board.
3. Request a review from the team, if not a draft.
4. Add targeted milestone, when applicable.
5. Create ticket tracking addition of public documentation pages entry, when applicable.
-->

JunDai commented:
@korshaknn / @Guardiola31337 : this is just the proposal of feedback caching change.
please help to review if this makes sense before I integrate it with the UI SDK.

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -662,6 +663,74 @@ class MapboxNavigation(
+     * @see [getCachedUserFeedback]
+     * @see [postCachedUserFeedback]
+     */
+    fun cacheUserFeedback(
```
why do we need both `cacheUserFeedback ()` and `postUserFeedback` ?
what usecase do we have for each of them?

JunDai replied:
in short, feedback 2.0 flow has different concept of 1tap and 2tap flow.
* for 1tap flow, user will use the `postUserFeedback` so the feedback will be queued to send.
* for 2tap flow, user will use the `cacheUserFeedback` so the feedback is cached and will be able to do update to it in a later time.

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -662,6 +663,74 @@ class MapboxNavigation(
+     *
+     * @return a list of [CachedNavigationFeedbackEvent]s
+     */
+    fun getCachedUserFeedback() = MapboxNavigationTelemetry.getCachedUserFeedback()
```
looks like we have the next flow
- cache events with `MapboxNavigationTelemetry`
- take them and add some details (what class is responsible for such changes?)
- use `MapboxNavigationTelemetry` to post events

is it a good idea to take events out of `MapboxNavigationTelemetry` to change them?
what happens if user closes the app/ cancel the session? with current implementation all pending events will be sent immediately. But looks like cached events will be lost

JunDai replied:
that's correct. 
* the flow you described is correct. That's why you can see I expose a new `CachedNavigationFeedbackEvent` instead of the internal `NavigationFeedbackEvent` and also only allow the `feedbackSubType` and `description` to be modified.
* yes, we take the risk of losing those cached events base on agreement of the feedback 2.0 flow discussion. I mentioned you in that ticket for reference.

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:
```diff
@@ -113,6 +112,7 @@ internal object MapboxNavigationTelemetry :
     private lateinit var locationsCollector: LocationsCollector
     private lateinit var sdkIdentifier: String
     private var logger: Logger? = null
+    private val feedbackEventCacheMap = LinkedHashMap<String, NavigationFeedbackEvent>()
```
does the order matters?

JunDai replied:
yes, we want to show the cached feedback events in the order of when it's created.

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:
```diff
@@ -159,7 +160,68 @@ internal object MapboxNavigationTelemetry :
+        val feedbackEventCache = LinkedHashMap(feedbackEventCacheMap)
+        feedbackEventCacheMap.clear()
+
+        cachedFeedbackEventList.forEach { cachedFeedback ->
+            metricsReporter.addEvent(
+                feedbackEventCache[cachedFeedback.feedbackId]?.apply {
+                    update(cachedFeedback)
+                } ?: return@forEach
+            )
```
why  do we need to update old events with new from a `cachedFeedbackEventList `?
can we just post new events because they are already updated?

JunDai replied:
when we post new events, we lose the location information. we only need the user's input of `feedbackSubType` and `description`, that's why I do this way.

korshaknn reviewed (commented):
we have some tailwork with Telemetry, it would be good to finish and test it first. And after that change/add something new.
and also we need to cover each new method with a bunch of test

JunDai commented:
> we have some tailwork with Telemetry, it would be good to finish and test it first. And after that change/add something new.
> and also we need to cover each new method with a bunch of test

yes, it completely makes sense. This draft is to give core team an idea about the proposal and make sure we can get an agreement of the new API. Also I will add test once the APIs are decided.

JunDai commented:
@korshaknn - when is the tailwork of Telemetry going to finish? and what do you think about the proposed apis?

korshaknn commented:
@JunDai thanks for your explanations! your changes LGTM! We still have some issues with telemetry, some of them might be on NN side. But I think it's not a blocker for your PR.
cc @Guardiola31337

JunDai commented:
@korshaknn / @Guardiola31337 : this PR is ready for reviewing, can you please take a look?
This is part of the Feedback 2.0 implementation. I would create another PR for the UI part so we can focus on our review on different areas.

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:
```diff
@@ -159,7 +162,68 @@ internal object MapboxNavigationTelemetry :
+        appMetadata: AppMetadata?
+    ) {
+        createUserFeedback(
+            {
+                sendMetricEvent(it)
+            },
```
let's move lambda to the last params position

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:
```diff
@@ -159,7 +162,68 @@ internal object MapboxNavigationTelemetry :
+        )
+    }
+
+    fun getCachedUserFeedback() = feedbackEventCacheMap.map {
```
return value for readability

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:
```diff
@@ -144,6 +146,7 @@ internal object MapboxNavigationTelemetry :
             "mapbox-navigation-android"
         }
         metricsReporter = reporter
+        feedbackEventCacheMap.clear()
```
we have a `private fun reset()` to reset data on `init` and on `sessionStop`.
do we need to move `feedbackEventCacheMap.clear()` to `reset()` ?
or it shouldn't be cleaned until new `initialize ` / `postCachedUserFeedback `?

JunDai replied:
> it shouldn't be cleaned until new initialize / postCachedUserFeedback 

that's correct.

korshaknn commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt:
```diff
@@ -342,6 +344,80 @@ class MapboxNavigationTelemetryTest {
+    @Test
+    fun cache_feedback_is_cached() {
+        baseMock()
+        mockAnotherRoute()
```
not needed

korshaknn commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt:
```diff
@@ -342,6 +344,80 @@ class MapboxNavigationTelemetryTest {
+        cacheUserFeedback(feedbackType = cacheFeedbackType)
+        val cacheFeedbackDescription = "cacheFeedbackDescription"
+        cacheUserFeedback(description = cacheFeedbackDescription)
+        updateSessionState(FREE_DRIVE)
```
not needed

JunDai replied:
if we don't add this line, I think we are not flushing the location buffer which in result will not call the cached user feedback event's `onBufferFull` callback.

korshaknn replied:
got it, you are right!

korshaknn commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetryTest.kt:
```diff
@@ -342,6 +344,80 @@ class MapboxNavigationTelemetryTest {
+        updateRouteProgress()
+        postUserFeedback()
+        cacheUserFeedback()
+        updateSessionState(FREE_DRIVE)
+        postCachedUserFeedback()
```
just to clarify
right now `cancel` event is always a final event in each session.
it means that on `cancel` event we call `locationsCollector.flushBuffers` and we immediately post all pending events (which are waiting for location buffer collection).

so, cachedEvents might be sent `after` `cancel`. 
@Guardiola31337

JunDai replied:
correct.

korshaknn reviewed (commented):
added some comments

JunDai commented:
@korshaknn / @Guardiola31337 - PR updated to address comments from latest review.

korshaknn reviewed (approved):
LGTM! 
@Guardiola31337 please take a look

Guardiola31337 commented on libnavigation-core/api/current.txt:
```diff
@@ -5,15 +5,18 @@ package com.mapbox.navigation.core {
     method public com.mapbox.navigation.core.trip.session.TripSessionState getTripSessionState();
     method public boolean navigateNextRouteLeg();
     method public void onDestroy();
-    method public void postUserFeedback(@com.mapbox.navigation.core.telemetry.events.FeedbackEvent.Type String feedbackType, String description, @com.mapbox.navigation.core.telemetry.events.FeedbackEvent.Source String feedbackSource, String? screenshot, String![]? feedbackSubType = emptyArray(), com.mapbox.navigation.core.telemetry.events.AppMetadata? appMetadata = null);
```
This is breaking SemVer. Notice `String![]? feedbackSubType = emptyArray()` vs `String![] feedbackSubType = emptyArray()`.

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -673,7 +674,7 @@ class MapboxNavigation(
         @FeedbackEvent.Source feedbackSource: String,
         screenshot: String?,
-        feedbackSubType: Array<String>? = emptyArray(),
+        feedbackSubType: Array<String> = emptyArray(),
```
See comment above re: breaking SemVer

JunDai commented:
@Guardiola31337 - would you mind to check again?
* I added one more test case `only_internal_cached_feedback_sent` to address the issue that we discussed offline.
* also reverted the change which causes the `SEMVER_MAJOR` concern.

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -683,6 +684,76 @@ class MapboxNavigation(
+    /**
+     * Create and cache a user feedback about an issue or problem with the Navigation SDK.
+     *
+     * Instead of sending the feedback, this API will cache the feedback in the Navigation SDK. The cached feedbacks can be retrieved by calling [getCachedUserFeedback] so the user can update and send them in a later time. To send the cached feedbacks, please use [postCachedUserFeedback].
```
We should be explicitly and add a note stating that this is the case only when in _Active Guidance_ (Feedback events are only handled in that mode) and as long as `postCachedUserFeedback` is called before transitioning between modes as that automatically will flush all the pending events. If not, all those events will be lost (as they're not push, only cached). BTW is that the behavior we want for such a use case? Have we tested this scenario? Wondering if we could be smarter when it comes to losing events 🤔 

Refs. https://github.com/mapbox/mapbox-navigation-android/pull/3724#discussion_r515205044

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:
```diff
@@ -174,6 +177,67 @@ internal object MapboxNavigationTelemetry :
+    }
+
+    fun postCachedUserFeedback(cachedFeedbackEventList: List<CachedNavigationFeedbackEvent>) {
+        log("post user feedback")
```
```suggestion
        log("post cached user feedback events")
```

Guardiola31337 reviewed (commented):
Thanks for addressing the feedback @JunDai 

Left a couple of minor comments / questions to discuss before merging here.

JunDai commented:
@Guardiola31337 - per @asinghal22 , we don't need to expose these APIs to external customer. so as we discussed offline, I moved them into an internal package. Also added more documents to those APIs per your suggestion. Please take a look again. Thanks!

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt:
```diff
@@ -0,0 +1,98 @@
+/**
+ * This object is only intent to be used internally and subject to change in future.
+ *
+ * This object is only a wrapper of [MapboxNavigationTelemetry] to cache/retrieve/post navigation.feedback event.
```
```suggestion
 * This object is only a wrapper of [MapboxNavigationTelemetry] to cache/retrieve/post a navigation.feedback event.
```

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt:
```diff
@@ -0,0 +1,98 @@
+ * It must be used after [MapboxNavigationTelemetry.initialize] has been called.
+ *
+ * A complete feedback cache flow will be:
+ * 1. [cacheUserFeedback] to generate navigation.feedback event in [MapboxNavigationTelemetry] internally.
```
```suggestion
 * 1. [cacheUserFeedback] to generate a navigation.feedback event in [MapboxNavigationTelemetry] internally.
```

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt:
```diff
@@ -0,0 +1,98 @@
+ * A complete feedback cache flow will be:
+ * 1. [cacheUserFeedback] to generate navigation.feedback event in [MapboxNavigationTelemetry] internally.
+ * 2. [getCachedUserFeedback] to retrieve all cached navigation.feedback events from [MapboxNavigationTelemetry].
+ * 3. make possible changes to the list that returned by [getCachedUserFeedback].
```
```suggestion
 * 3. make possible changes to the list that is returned by [getCachedUserFeedback].
```

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/MapboxNavigationFeedbackCache.kt:
```diff
@@ -0,0 +1,98 @@
+     * * Retrieve the cached feedback list by calling [getCachedUserFeedback].
+     * * Update the cached feedback list to provide more information of the feedback.
+     * * Call [postCachedUserFeedback] to send the updated cached feedback list.
+     *
```
What about adding the note from above here too?

```kotlin
 * Note:
 * 1. Only the feedback events created by [cacheUserFeedback] and returned by [getCachedUserFeedback]
 * are valid in [postCachedUserFeedback]. All user manually created events will be dropped.
 * 2. The cache APIs should only be used when in Active Guidance mode since navigation.feedback events are only handled in that mode.
 * 3. [postCachedUserFeedback] should be called before transitioning between modes. Otherwise, all cached events will be lost.
```

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/events/CachedNavigationFeedbackEvent.kt:
```diff
@@ -0,0 +1,20 @@
+ * @param screenshot encoded screenshot (optional)
+ * @param feedbackSubType array of [FeedbackEvent.Description] (optional)
+ */
+class CachedNavigationFeedbackEvent internal constructor(
```
`CachedNavigationFeedbackEvent` can be `internal` or moved into the `internal` package otherwise.

JunDai replied:
this event should be exposed, because the UI side will need to use it. Please refer to the Feedback UX PR.

Guardiola31337 replied:
It can be "public" but under the `internal` package so it'll be accessible from UI side.

> or moved into the `internal` package

Guardiola31337 reviewed (commented):
Left some minor comments @JunDai

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/telemetry/CachedNavigationFeedbackEvent.kt:
```diff
@@ -0,0 +1,53 @@
+ * @param screenshot encoded screenshot (optional)
+ * @param feedbackSubType array of [FeedbackEvent.Description] (optional)
+ */
+class CachedNavigationFeedbackEvent internal constructor(
```
I believe this can be now a `data class` instead of a `Builder`. BTW if we want to leave it as is, why are we missing the `toString`? cc @LukasPaczos

LukasPaczos replied:
`data class` exposes a public `.copy()` method. If someone relies on that method, their build will break when we add a new constructor parameter, even if the constructor itself is `internal`.

Guardiola31337 replied:
The class itself is `internal` so that shouldn't be a problem, but as I mentioned I'm fine with `Builder` too, as long as we add the `toString` 😅

LukasPaczos replied:
Oh I didn't notice you moved it to `internal` package :+1: if it was public, it wouldn't have to have a builder pattern anyway, just a `class CachedNavigationFeedbackEvent internal constructor` with `equals/hash/toString` overrides.

Guardiola31337 replied:
Yeah, that's correct.

Guardiola31337 replied:
I meant "builder" 😅

Guardiola31337 reviewed (approved):
Thanks for addressing the feedback @JunDai :shipit:
